lisa/nic.py

```diff
@@ -402,57 +348,128 @@ def _get_node_nic_info(self) -> None:
                 expected_exit_code_failure_message="Could not grab NIC device info.",
             )
 
+        has_pci_module = False
         for line in result.stdout.splitlines():
-            sriov_match = self.__nic_lower_regex.search(line)
+            sriov_match = self.__nic_pci_device_regex.search(line)
             if sriov_match:
-                upper_nic, lower_nic, pci_slot = sriov_match.groups()
-                nic_info = NicInfo(upper_nic, lower_nic, pci_slot)
-                self.append(nic_info)
+                nic_name, pci_device_name, pci_slot = sriov_match.groups()
+                used_module = lspci.get_used_module(pci_slot)
+                if used_module:
+                    has_pci_module = True
+                self.append(
+                    NicInfo(
+                        name=nic_name,
+                        pci_device_name=pci_device_name,
+                        has_pci_device=True,
+                        has_pci_module=has_pci_module,
+                        pci_slot=pci_slot,
+                        pci_module_name=used_module,
+                    )
+                )
+
             sriov_match = self.__nic_vf_slot_regex.search(line)
             if sriov_match:
-                lower_nic, pci_slot = sriov_match.groups()
+                pci_device_name, pci_slot = sriov_match.groups()
                 ip = self._node.tools[Ip]
-                lower_nic_mac = ip.get_mac(lower_nic)
-                for nic_name in [x for x in self.nic_names if x != lower_nic]:
-                    upper_nic_mac = ip.get_mac(nic_name)
-                    if upper_nic_mac == lower_nic_mac:
-                        upper_nic = nic_name
-                        nic_info = NicInfo(upper_nic, lower_nic, pci_slot)
-                        self.append(nic_info)
+                pci_nic_mac = ip.get_mac(pci_device_name)
+                for nic_name in [x for x in self._nic_names if x != pci_device_name]:
+                    synthetic_nic_mac = ip.get_mac(nic_name)
+                    if synthetic_nic_mac == pci_nic_mac:
+                        used_module = lspci.get_used_module(pci_slot)
+                        if used_module:
+                            has_pci_module = True
+                        self.append(
+                            NicInfo(
+                                name=nic_name,
+                                pci_device_name=pci_device_name,
+                                has_pci_device=True,
+                                has_pci_module=has_pci_module,
+                                pci_slot=pci_slot,
+                                pci_module_name=used_module,
+                            )
+                        )
```

Comment:
Should I initialize the self.nics to store all the nics? then use different method or variable to get the upper/lower nics etc?

Reply:
Let's consider it from two ways.

1. Can it meet all current and future known scenarios? If yes, we keep the current design.
2. From the names, we should use lower to save like "eth1" instead of pci_device_name. It looks clearer than "pci_device_name". Because in no synthetic scenario, the name and pci_device_name should be the same. But it's hard to use.
3. How about add a method called like `get_pci_device_names`. It can returns all `name` without synthetic AND `lower` of synthetic nics.